### PR TITLE
src: include sync stop for heap object tracking

### PR DESCRIPTION
**Pull request by juanarbol**
This patch includes `StopTrackingHeapObjects` synchronously. NSolid will
grab all in-progress profiles on exit.

Also, this path will convert the tracker stopper serialization async.

PR-URL: https://github.com/nodesource/nsolid/pull/92

### CODE REVIEW

File: src/nsolid.h [hunk 1]
```diff
@@ -1005,6 +1005,16 @@ class NODE_EXTERN Snapshot {
    */
   static int StopTrackingHeapObjects(SharedEnvInst envinst);
 
+  /**
+   * @brief same as TakeSnapshot but stops tracking heap objects in the heap
+   * profiler synchronously.
```
Comment (santigimeno) [inline src/nsolid.h:1010]: Seems TakeSnapshot is a leftover

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 2]
```diff
@@ -83,6 +83,71 @@ int NSolidHeapSnapshot::StopTrackingHeapObjects(SharedEnvInst envinst) {
...
+  if (snapshot == nullptr) {
+    stor.cb(heap_profiler::HEAP_SNAPSHOT_FAILURE,
+            std::string(),
+            stor.data.get());
+  } else {
+    std::string snapshot_str;
+    DataOutputStream<uint64_t, v8::HeapSnapshot> stream(&snapshot_str,
+                                                        snapshot,
+                                                        &thread_id);
+
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: Why can't we use the "streaming" version of the DataOutputStream here. Both are sync so we should be fine

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 3]
```diff
@@ -83,6 +83,71 @@ int NSolidHeapSnapshot::StopTrackingHeapObjects(SharedEnvInst envinst) {
...
+    stor.is_tracking_heapobjects_ = false;
+
+
+    // At this point, the snapshot is fully serialized
+    stor.cb(0, snapshot_str.c_str(), stor.data.get());
+    // Tell ZMQ that the snapshot is done
+    stor.cb(0, std::string(), stor.data.get());
+
+    // Don't leak the snapshot string
+    snapshot_str.clear();
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: I don't understand why we're doing this

**Review by santigimeno — CHANGES_REQUESTED** [review]
Looks good: I'll test in a few minutes

**Review by santigimeno — CHANGES_REQUESTED** [review]
Just tested, everything's good. I'll approve once the nits are addressed

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 2 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: This comment as discussed is for the async stoped. Fixed in https://github.com/nodesource/nsolid/commit/37d63c035e309c433c4af1c02774d4aec1f79ee5

**Review by trevnorris — APPROVED** [review]
Everything now passes with the additional changes.

**Review by santigimeno — APPROVED** [review]

**Pull request merged**